### PR TITLE
py-fairscale and dependencies py-pgzip libgit2 py-pygit2

### PR DESCRIPTION
**Pull request by sidpbury**
This pull request for py-fairscale and its dependencies: py-fairscale and py-pgzip are new.  libgit2 and py-pygit2 are updates to existing packages.

### CODE REVIEW

**Review by tldahlgren — COMMENTED** [review]
Confirmed the home pages of the new packages and all  of the new version sha256s.

File: var/spack/repos/builtin/packages/py-fairscale/package.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+
+class PyFairscale(PythonPackage):
+    """FairScale is a PyTorch extension library for high performance and large
+    scale training. This library extends basic PyTorch capabilities while adding
+    new SOTA scaling techniques."""
+
+    homepage = "https://github.com/facebookresearch/fairscale"
+
+    url = "https://github.com/facebookresearch/fairscale/archive/refs/tags/v0.4.13.zip"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fairscale/package.py]: https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi-vs-github

File: var/spack/repos/builtin/packages/py-fairscale/package.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    version("0.4.13", sha256="a6012a6f23eb0e80dc209dfea992336b4d0d8a191a9080cb746a2b78e5b76cd1")
+    version("0.4.4", sha256="b4f9a9ad82e562a1d19ad9249238d135395bc8f4068d292f639308951efff5a8")
+
+    variant(
+        "dev-tools", default=False, description="support for cuda.list.gpu, mypy, scaler, weight"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fairscale/package.py]: I would call the variant `+dev` to match the file name.

File: var/spack/repos/builtin/packages/py-fairscale/package.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    depends_on("python@3.8:", type=("build", "run"))
+    depends_on("ninja", type="build")
+    # from pyproject.toml
+    depends_on("py-wheel@0.30.0:", type="build")
+    depends_on("py-setuptools@40.6.2:", type="build")
+    # from requirements.txt
+    depends_on("py-torch@1.8.0:", type=("build", "run"))
+    depends_on("py-numpy@1.22.0:", type=("build", "run"))
+    # from requirements-dev.txt
+    with when("+dev-tools"):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fairscale/package.py]: Seems like you're missing a lot of deps. What's the motivation for adding the variant? `pip install fairscale[dev]` isn't even an option

File: var/spack/repos/builtin/packages/py-pgzip/package.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+class PyPgzip(PythonPackage):
+    """A multi-threading implementation of Python gzip module"""
+
+    homepage = "https://github.com/pgzip/pgzip"
+    pypi = "pgzip/pgzip-0.3.4.tar.gz"
+
+    version("0.3.4", sha256="ef56449039bc6e88558e46fe6bb11e3faaeef445d3985a9fb286795ff842c480")
+    version("0.3.1", sha256="a9c2df369311473ec3c239f26bf01638bdc6b6094d89ff4c81c6ef5c84eb24b7")
+
+    depends_on("python", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pgzip/package.py]: This is already added by the base class and can be removed

File: var/spack/repos/builtin/packages/py-pygit2/package.py [hunk 5]
```diff
@@ -23,17 +25,21 @@ class PyPygit2(PythonPackage):
...
     depends_on("libgit2@0.28.0:0.28", when="@0.28:1.0")
     depends_on("libgit2@0.24:0.27", when="@0.24:0.27")
-    depends_on("python@3.6:", when="@1.4.0:")
-    depends_on("python@3.6:3.8", when="@1.2:1.3")
-    depends_on("python@3.5:3.8", when="@1.0:1.1")
-    depends_on("python@2.7:3.7", when="@0.28:0")
+    depends_on("python@3.8:", when="@1.11.1")
+    depends_on("python@3.7:", when="@1.4.0:1.6.0")
+    depends_on("python@:3.8", when="@1.2:1.3")
+    depends_on("python@:3.8", when="@1.0:1.1")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygit2/package.py]:
```suggestion
    depends_on("python@:3.8", when="@1.0:1.3")
```

File: var/spack/repos/builtin/packages/py-pygit2/package.py [hunk 6]
```diff
@@ -23,17 +25,21 @@ class PyPygit2(PythonPackage):
     depends_on("py-setuptools", type="build")
     # Version must match with libgit2
     # See: https://www.pygit2.org/install.html
-    depends_on("libgit2@1.1.0:1.1", when="@1.4:")
+    depends_on("libgit2@1.6.0:1.6", when="@1.12.0:1.12")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygit2/package.py]:
```suggestion
    depends_on("libgit2@1.6", when="@1.12")
```
Same can be used for other versions

File: var/spack/repos/builtin/packages/py-fairscale/package.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    depends_on("ninja", type="build")
+    # from pyproject.toml
+    depends_on("py-wheel@0.30.0:", type="build")
+    depends_on("py-setuptools@40.6.2:", type="build")
+    # from requirements.txt
+    depends_on("py-torch@1.8.0:", type=("build", "run"))
+    depends_on("py-numpy@1.22.0:", type=("build", "run"))
+    # added extra to support cuda.list.gpu, scaler, and weight (not in pip install)
+    with when("extra"):
+        depends_on("py-pynvml@8.0.4", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fairscale/package.py:34]: This doesn't seem to be imported anywhere in the source code

File: var/spack/repos/builtin/packages/py-fairscale/package.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        depends_on("py-numpy@1.22.0:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fairscale/package.py]: This is already always required

File: var/spack/repos/builtin/packages/py-fairscale/package.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        depends_on("py-scikit-learn@1.1.3", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fairscale/package.py:35]: Doesn't seem to be imported

File: var/spack/repos/builtin/packages/py-fairscale/package.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        depends_on("py-pygit2@1.11.1", type=("build", "run"))
+        depends_on("py-pgzip@0.3.1", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fairscale/package.py:37]: Are these documented anywhere? Can you add a comment so it's easy to find the required versions?

File: var/spack/repos/builtin/packages/py-pygit2/package.py [hunk 11]
```diff
@@ -12,8 +12,10 @@ class PyPygit2(PythonPackage):
...
 
     homepage = "https://www.pygit2.org/"
-    pypi = "pygit2/pygit2-0.24.1.tar.gz"
+    pypi = "pygit2/pygit2-1.12.2.tar.gz"
 
+    version("1.12.2", sha256="56e85d0e66de957d599d1efb2409d39afeefd8f01009bfda0796b42a4b678358")
+    version("1.11.1", sha256="793f583fd33620f0ac38376db0f57768ef2922b89b459e75b1ac440377eb64ec")
     version("1.6.0", sha256="7aacea4e57011777f4774421228e5d0ddb9a6ddb87ac4b542346d17ab12a4d62")
     version("1.4.0", sha256="cbeb38ab1df9b5d8896548a11e63aae8a064763ab5f1eabe4475e6b8a78ee1c8")
     version("1.3.0", sha256="0be93f6a8d7cbf0cc79ae2f0afb1993fc055fc0018c27e2bd01ba143e51d4452")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygit2/package.py:21]: You should probably remove 0.24.1, it requires versions of Python and libgit2 that we don't have packaged.

File: var/spack/repos/builtin/packages/py-pygit2/package.py [hunk 12]
```diff
@@ -23,17 +25,20 @@ class PyPygit2(PythonPackage):
     depends_on("py-setuptools", type="build")
     # Version must match with libgit2
     # See: https://www.pygit2.org/install.html
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygit2/package.py]:
```suggestion
    # https://www.pygit2.org/install.html#version-numbers
```

File: var/spack/repos/builtin/packages/py-pygit2/package.py [hunk 13]
```diff
@@ -23,17 +25,20 @@ class PyPygit2(PythonPackage):
...
+    depends_on("libgit2@1.5", when="@1.11")
+    depends_on("libgit2@1.1", when="@1.4")
+    depends_on("libgit2@1.0", when="@1.2:1.3")
+    depends_on("libgit2@:1.0", when="@1.1")
+    depends_on("libgit2@:0.28", when="@:1.0")
+    depends_on("libgit2@:0.27", when="@:0.27")
+    depends_on("python@3.8:", when="@1.11.1")
+    depends_on("python@3.7:", when="@1.4.0:1.6.0")
+    depends_on("python@:3.8", when="@1.0:1.3")
+    depends_on("python@3.7", when="@0.28:0")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygit2/package.py]:
```suggestion
    depends_on("libgit2@1.6", when="@1.12")
    depends_on("libgit2@1.5", when="@1.10:1.11")
    depends_on("libgit2@1.4", when="@1.9")
    depends_on("libgit2@1.3", when="@1.7:1.8")
    depends_on("libgit2@1.1", when="@1.4:1.6")
    depends_on("libgit2@1.0", when="@1.2:1.3")
    depends_on("libgit2@0.99:1.0", when="@1.1")
    depends_on("libgit2@0.28", when="@0.28:1.0")
    depends_on("python@3.8:3.11", when="@1.11:")
    depends_on("python@:3.10", when="@1.7:1.10")
    depends_on("python@:3.9", when="@1.4:1.6")
    depends_on("python@:3.8", when="@1.0:1.3")
    depends_on("python@:3.7", when="@0.28")
```

File: var/spack/repos/builtin/packages/py-pygit2/package.py [hunk 14]
```diff
@@ -23,17 +25,20 @@ class PyPygit2(PythonPackage):
...
     depends_on("py-six", type=("build", "run"), when="@:0.28.2")
-    depends_on("py-cffi@1.4.0:", type=("build", "run"))
+    depends_on("py-cffi@1.9.1:", when="@1.11.1:", type=("build", "run"))
+    depends_on("py-cffi@1.4.0:", when="@:1.6.0", type=("build", "run"))
     depends_on("py-cached-property", when="@1.1.0:1.5", type=("build", "run"))
     depends_on("py-cached-property", when="@1.6.0: ^python@:3.7", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygit2/package.py]: Seems like this dep is no longer required

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-fairscale/package.py (hunk 7 above)
Comment (sidpbury) [reply var/spack/repos/builtin/packages/py-fairscale/package.py:34]: added a commit on this

File: var/spack/repos/builtin/packages/py-fairscale/package.py (hunk 8 above)
Comment (sidpbury) [reply var/spack/repos/builtin/packages/py-fairscale/package.py]: removed as suggested

File: var/spack/repos/builtin/packages/py-fairscale/package.py (hunk 10 above)
Comment (sidpbury) [reply var/spack/repos/builtin/packages/py-fairscale/package.py:37]: added a commit

File: var/spack/repos/builtin/packages/py-fairscale/package.py (hunk 9 above)
Comment (sidpbury) [reply var/spack/repos/builtin/packages/py-fairscale/package.py:35]: +extra added

File: var/spack/repos/builtin/packages/py-pygit2/package.py (hunk 11 above)
Comment (sidpbury) [reply var/spack/repos/builtin/packages/py-pygit2/package.py:21]: removed 0.24.1

File: var/spack/repos/builtin/packages/py-pygit2/package.py (hunk 12 above)
Comment (sidpbury) [reply var/spack/repos/builtin/packages/py-pygit2/package.py]: made the change

File: var/spack/repos/builtin/packages/py-pygit2/package.py (hunk 13 above)
Comment (sidpbury) [reply var/spack/repos/builtin/packages/py-pygit2/package.py]: change made

File: var/spack/repos/builtin/packages/py-pygit2/package.py (hunk 14 above)
Comment (sidpbury) [reply var/spack/repos/builtin/packages/py-pygit2/package.py]: change made

File: var/spack/repos/builtin/packages/py-fairscale/package.py (hunk 7 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-fairscale/package.py:34]: Did you forget to push the commit? The dep is still listed but still not imported

File: var/spack/repos/builtin/packages/py-pygit2/package.py (hunk 14 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pygit2/package.py]: Did you forget to push the commit? The package still lists cached-property as required

Comment (sidpbury) [reply var/spack/repos/builtin/packages/py-pygit2/package.py]: I am not sure what happened; I just did another push.  There should not have been any more commits, but there seemed to be.  The screen I am on right now is too small to see anything.  When I get to a larger screen, I will check what those commits were. I hope I didn't create a mess.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pygit2/package.py]: Doesn't appear to be any changes to py-cached-property

**Comment by sidpbury** [comment]
spack style --fix

File: var/spack/repos/builtin/packages/py-pygit2/package.py [hunk 15]
```diff
@@ -12,30 +12,36 @@ class PyPygit2(PythonPackage):
...
-    depends_on("libgit2@1.1.0:1.1", when="@1.4:")
-    depends_on("libgit2@1.0.0:1.0", when="@1.2:1.3")
-    depends_on("libgit2@0:1.0", when="@1.1.0:1.1")
-    depends_on("libgit2@0.28.0:0.28", when="@0.28:1.0")
-    depends_on("libgit2@0.24:0.27", when="@0.24:0.27")
-    depends_on("python@3.6:", when="@1.4.0:")
-    depends_on("python@3.6:3.8", when="@1.2:1.3")
-    depends_on("python@3.5:3.8", when="@1.0:1.1")
-    depends_on("python@2.7:3.7", when="@0.28:0")
+    depends_on("py-wheel", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygit2/package.py]: Wheel is already added by the base class, you don't need to add it again

File: var/spack/repos/builtin/packages/py-pygit2/package.py [hunk 16]
```diff
@@ -12,30 +12,36 @@ class PyPygit2(PythonPackage):
...
+    depends_on("python@:3.8", when="@1.0:1.3")
+    depends_on("python@:3.7", when="@0.28")
     depends_on("py-six", type=("build", "run"), when="@:0.28.2")
-    depends_on("py-cffi@1.4.0:", type=("build", "run"))
-    depends_on("py-cached-property", when="@1.1.0:1.5", type=("build", "run"))
-    depends_on("py-cached-property", when="@1.6.0: ^python@:3.7", type=("build", "run"))
+    depends_on("py-cffi@1.4.0:", when="@:1.5", type=("build", "run"))
+    depends_on("py-cffi@1.6.0:", when="@1.6:1.7", type=("build", "run"))
+    depends_on("py-cffi@1.9.1:", when="@1.8:", type=("build", "run"))
+    depends_on("py-cached-property", when="@1.1:1.9", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygit2/package.py]: Not required for 1.6 when Python 3.8+, the previous condition was correct

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-pygit2/package.py [hunk 17]
```diff
@@ -12,30 +12,35 @@ class PyPygit2(PythonPackage):
...
+    depends_on("python@:3.8", when="@1.0:1.3")
+    depends_on("python@:3.7", when="@0.28")
     depends_on("py-six", type=("build", "run"), when="@:0.28.2")
-    depends_on("py-cffi@1.4.0:", type=("build", "run"))
-    depends_on("py-cached-property", when="@1.1.0:1.5", type=("build", "run"))
-    depends_on("py-cached-property", when="@1.6.0: ^python@:3.7", type=("build", "run"))
+    depends_on("py-cffi@1.4.0:", when="@:1.5", type=("build", "run"))
+    depends_on("py-cffi@1.6.0:", when="@1.6:1.7", type=("build", "run"))
+    depends_on("py-cffi@1.9.1:", when="@1.8:", type=("build", "run"))
+    depends_on("py-cached-property", when="@1.1:1.9 ^python@:3.7", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygit2/package.py]:
```suggestion
    depends_on("py-cached-property", when="@1.1:1.5", type=("build", "run"))
    depends_on("py-cached-property", when="@1.6: ^python@:3.7", type=("build", "run"))```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-pygit2/package.py (hunk 17 above)
Comment (sidpbury) [reply var/spack/repos/builtin/packages/py-pygit2/package.py]: changes made as requested

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**